server/src/main/java/org/candlepin/model/PoolCurator.java

```diff
@@ -2006,4 +2006,217 @@ public void clearPoolSourceEntitlementRefs(Iterable<String> poolIds) {
 
         return output;
     }
+
+    /**
+     * Fetches a mapping of pool IDs to sets of product UUIDs representing the provided products of
+     * the given pool. The returned map will only contain mappings for pools specified in the given
+     * collection of pool IDs.
+     *
+     * @param poolIds
+     *  A collection of pool IDs for which to fetch provided product UUIDS
+     *
+     * @return
+     *  A mapping of pool IDs to provided product UUIDs
+     */
+    public Map<String, Set<String>> getProvidedProductUuidsByPoolIds(Collection<String> poolIds) {
+        Map<String, Set<String>> providedProductMap = new HashMap<>();
+
+        if (poolIds != null && !poolIds.isEmpty()) {
+            StringBuilder builder =
+                new StringBuilder("SELECT p.id, pp.uuid FROM Pool p JOIN p.providedProducts pp WHERE");
+            javax.persistence.Query query = null;
+
+            int blockSize = getInBlockSize();
+            int blockCount = (int) Math.ceil(poolIds.size() / (float) blockSize);
+
+            if (blockCount > 1) {
+                Iterable<List<String>> blocks = Iterables.partition(poolIds, blockSize);
+
+                for (int i = 0; i < blockCount; ++i) {
+                    if (i != 0) {
+                        builder.append(" OR");
+                    }
+
+                    builder.append(" p.id IN (:block").append(i).append(')');
+                }
+
+                query = this.getEntityManager().createQuery(builder.toString());
+                int i = -1;
+
+                for (List<String> block : blocks) {
+                    query.setParameter("block" + ++i, block);
+                }
+            }
+            else {
+                builder.append(" p.id IN (:pids)");
+                query = this.getEntityManager().createQuery(builder.toString())
+                    .setParameter("pids", poolIds);
+            }
+
+            for (Object[] cols : (List<Object[]>) query.getResultList()) {
+                Set<String> providedProducts = providedProductMap.get((String) cols[0]);
+
+                if (providedProducts == null) {
+                    providedProducts = new HashSet<>();
+                    providedProductMap.put((String) cols[0], providedProducts);
+                }
+
+                providedProducts.add((String) cols[1]);
+            }
```

Comment:
This logic is repeated across multiple methods. We should refactor this into reusable pieces.

Reply:
Refactored the logic.